### PR TITLE
Changed gRPC to generic transport in Ordering Service

### PR DESCRIPTION
**Pull request by x3medima17**
Changed gRPC to generic transport in Ordering Service

### CODE REVIEW

File: irohad/ordering/CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +15,8 @@
 add_library(ordering_service
     impl/ordering_gate_impl.cpp
     impl/ordering_service_impl.cpp
-    )
+    impl/ordering_gate_transport_grpc.cpp impl/ordering_gate_transport_grpc.hpp
+        impl/ordering_service_transport_grpc.cpp impl/ordering_service_transport_grpc.hpp)
```
Comment (lebdron) [inline irohad/ordering/CMakeLists.txt]: Remove headers, fix indent.

File: irohad/network/ordering_service_transport.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#define IROHA_ORDERING_SERVICE_TRANSPORT_H
+
+#include "ordering.grpc.pb.h"
+#include "model/proposal.hpp"
+#include "ordering.grpc.pb.h"
+
+namespace iroha {
+  namespace network {
+
+    class OrderingServiceTransport {
```
Comment (lebdron) [inline irohad/network/ordering_service_transport.hpp]: Add description to class and methods.

File: irohad/network/ordering_gate_transport.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    class OrderingGateNotification {
+    public:
+      virtual void handleProposal(std::shared_ptr<iroha::model::Proposal>) = 0;
+
+      virtual ~OrderingGateNotification() = default;
+    };
+
+
+    class OrderingGateTransport {
```
Comment (lebdron) [inline irohad/network/ordering_gate_transport.hpp]: Add description to class and methods. What is subscriber?

File: irohad/network/ordering_gate_transport.hpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#include "model/peer.hpp"
+#include "model/transaction.hpp"
+#include "model/proposal.hpp"
+#include "ordering.grpc.pb.h"
+
+
+namespace iroha {
+  namespace network {
+
+    class OrderingGateNotification {
```
Comment (lebdron) [inline irohad/network/ordering_gate_transport.hpp]: Add description to class and methods. What is the purpose of this interface?

File: irohad/network/ordering_gate_transport.hpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    class OrderingGateTransport {
+
+    public:
+      std::shared_ptr<OrderingGateNotification> subscriber_ {nullptr};
+
+      virtual void propagate(std::shared_ptr<const iroha::model::Transaction>) = 0;
+
+      virtual void subscribe(std::shared_ptr<OrderingGateNotification>) = 0;
+
+      virtual std::shared_ptr<model::Proposal> getProposal(const iroha::ordering::proto::Proposal *request) = 0;
```
Comment (lebdron) [inline irohad/network/ordering_gate_transport.hpp]: Probably protobuf is gRPC-specfic, so it should only appear in `OrderingGateTransportGrpc`.

File: irohad/network/ordering_service_transport.hpp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef IROHA_ORDERING_SERVICE_TRANSPORT_H
+#define IROHA_ORDERING_SERVICE_TRANSPORT_H
+
+#include "ordering.grpc.pb.h"
+#include "model/proposal.hpp"
+#include "ordering.grpc.pb.h"
```
Comment (lebdron) [inline irohad/network/ordering_service_transport.hpp]: Duplicate include of `ordering.grpc.pb.h`.

File: irohad/ordering/impl/ordering_gate_impl.cpp [hunk 7]
```diff
@@ -53,15 +47,16 @@ namespace iroha {
...
       log_->info("transactions in proposal: {}", transactions.size());
 
-      model::Proposal proposal(transactions);
-      proposal.height = request->height();
-      handleProposal(std::move(proposal));
+      auto proposal = transport_->getProposal(request);
 
+      proposal->height = request->height();
+      transport_->subscriber_->handleProposal(proposal);
+      //handleProposal(proposal);
```
Comment (lebdron) [inline irohad/ordering/impl/ordering_gate_impl.cpp]: Why does the received proposal go to transport again?

File: irohad/ordering/impl/ordering_gate_impl.cpp [hunk 8]
```diff
@@ -53,15 +47,16 @@ namespace iroha {
...
       return grpc::Status::OK;
     }
 
-    void OrderingGateImpl::handleProposal(model::Proposal &&proposal) {
-      proposals_.get_subscriber().on_next(proposal);
+    void OrderingGateImpl::handleProposal(std::shared_ptr<model::Proposal> proposal) {
```
Comment (lebdron) [inline irohad/ordering/impl/ordering_gate_impl.cpp]: This proposal is only used in this method, so why use `shared_ptr`?

File: irohad/ordering/impl/ordering_gate_transport_grpc.hpp [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+      OrderingGateTransportgRPC& operator=(const OrderingGateTransportgRPC&& rhs) = delete;
+
+      void propagate(std::shared_ptr<const iroha::model::Transaction> transaction)  override;
+
+      void subscribe(std::shared_ptr<iroha::network::OrderingGateNotification> subscriber) override;
+
+
+      std::shared_ptr<model::Proposal> getProposal(const iroha::ordering::proto::Proposal *request) override;
+
+      ~OrderingGateTransportgRPC() override = default;
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply irohad/ordering/impl/ordering_gate_transport_grpc.hpp]: Mistake, already deleted

File: irohad/main/impl/ordering_init.cpp [hunk 10]
```diff
@@ -16,21 +16,27 @@
...
 
 #include "main/impl/ordering_init.hpp"
+#include "ordering/impl/ordering_gate_transport_grpc.hpp"
+#include "ordering/impl/ordering_service_transport_grpc.hpp"
 
 namespace iroha {
   namespace network {
     auto OrderingInit::createGate(std::string network_address) {
-      return std::make_shared<ordering::OrderingGateImpl>(network_address);
+      auto transport = std::make_shared<iroha::ordering::OrderingGateTransportgRPC>(network_address);
```
[the remark this answers is not in the document]
Comment (x3medima17) [reply irohad/main/impl/ordering_init.cpp]: I did not write this method, should I still chage its name?

File: irohad/ordering/impl/ordering_gate_impl.cpp (hunk 7 above)
Comment (x3medima17) [reply irohad/ordering/impl/ordering_gate_impl.cpp]: first it gets a proposal from request, then it handles it

File: irohad/ordering/impl/ordering_gate_impl.cpp (hunk 8 above)
Comment (x3medima17) [reply irohad/ordering/impl/ordering_gate_impl.cpp]: I talked with Fyodor and we decidet to use shared_ptr in order to be safe from seg faults, if client destructs object in another thread

**Pull request closed without merging**